OpenRA.Mods.Common/Activities/Resupply.cs

```diff
@@ -78,18 +78,29 @@ public override bool Tick(Actor self)
 				if (aircraft != null)
 				{
 					aircraft.AllowYieldingReservation();
-					if (aircraft.Info.FlightDynamics.HasFlag(FlightDynamic.TakeOffOnResupply))
-						Queue(new TakeOff(self));
 
-					return true;
+					// ReturnToBase only handles TakeOffOnResupply when Resupply finishes, not when it's cancelled prematurely
+					if (aircraft.Info.FlightDynamics.HasFlag(FlightDynamic.TakeOffOnResupply))
+						QueueChild(new TakeOff(self));
 				}
-				else if (repairableNear != null)
-					return true;
 				else
 				{
-					QueueChild(move.MoveToTarget(self, host));
-					return false;
+					// If there's no next activity, move to rallypoint if available, else just leave host if Repairable.
+					// Do nothing if RepairableNear (RepairableNear actors don't enter their host and will likely remain within closeEnough).
+					// If there's a next activity and we're not RepairableNear, first leave host if the next activity is not a Move.
+					if (self.CurrentActivity.NextActivity == null)
```

Comment:
It would be nice if ground units also had the option to stay on the resupplier after resupplying. This could be done by replacing `TakeOffOnResupply` with `LeaveOnRepair` and `LeaveOnRearm` which would be parameters of the `Repairable` and `Rearmable` trait respectively. This would be a step in separating the resupply logic from `Aircraft`. Having separate booleans for repair and rearm has the added benefit of allowing aircraft to takeoff after repairing on the repair pad but not after rearming on the airfield.

Reply:
I agree, but I'd like to leave that step for a follow-up, we're already quite late with the playtest.

Reply:
Yes, this is more something for a follow-up than something that needs to happen right now.